scripts/check_compliance.py

```diff
@@ -426,73 +426,123 @@ class Codeowners(ComplianceTest):
     _name = "Codeowners"
     _doc  = "https://help.github.com/articles/about-code-owners/"
 
-    def parse_codeowners(self, git_root, codeowners):
-        all_files = []
+
+    def ls_owned_files(self, codeowners):
+        """Returns a OrderedDict mapping git patterns from the CODEOWNERS file
+        to the corresponding list of files found on the filesystem.  It
+        unfortunately does not seem possible to invoke git and re-use
+        how 'git ignore' and/or 'git attributes' already implement this,
+        we must re-invent it.
+        """
+
+        # TODO: filter out files not in "git ls-files" (e.g.,
+        # sanity-out) _if_ the overhead isn't too high for a clean tree.
+        #
+        # pathlib.match() doesn't support **, so it looks like we can't
+        # recursively glob the output of ls-files directly, only real
+        # files :-(
+
+        pattern2files = collections.OrderedDict()
+        top_path = Path(git_top)
+        top_len = len(git_top) + 1 # 1 for the slash
+
         with open(codeowners, "r") as codeo:
-            for line in codeo.readlines():
-                if not line.startswith("#") and line != "\n":
-                    match = re.match(r"([^\s]+)\s+(.*)", line)
-                    if match:
-                        add_base = False
-                        path = match.group(1)
-                        if path.startswith("/"):
-                            abs_path = git_root + path
-                        else:
-                            abs_path = "**/{}".format(path)
-                            add_base = True
-
-                        if abs_path.endswith("/"):
-                            abs_path = abs_path + "**"
-                        elif os.path.isdir(abs_path):
-                            self.add_failure("Expected / after directory '{}' "
-                                             "in CODEOWNERS".format(path))
-                            continue
-                        g = glob.glob(abs_path, recursive=True)
-                        if not g:
-                            self.add_failure("Path '{}' not found, in "
-                                             "CODEOWNERS".format(path))
-                        else:
-                            files = []
-                            if not add_base:
-                                for f in g:
-                                    l = f.replace(git_root + "/", "")
-                                    files.append(l)
-                            else:
-                                files = g
-
-                            all_files += files
-
-        files = []
-        for f in all_files:
-            if os.path.isfile(f):
-                files.append(f)
-
-        return set(files)
+            linenum = 0
+            for line in codeo:
```

Comment:
Can do `for lineno, line in enumerate(codeo, 1):`.

(I used `linenr` in Kconfiglib, but later on realized that `lineno` is the "standard" in a bunch of other places, so just to avoid repeating that. :P)